File: python/adb_auto_player/ipc/game_gui.py
"""ADB Auto Player Game GUI Module."""

from dataclasses import dataclass
from typing import Any

from .constraint import ConstraintType
from .menu_option import MenuOption


@dataclass
class GameGUIOptions:
    """Game GUI Options."""

    game_title: str
    menu_options: list[MenuOption]
    categories: list[str]
    config_path: str | None = None
    constraints: dict[str, dict[str, ConstraintType]] | None = None

    def to_dict(self):
        """Convert to dict for JSON serialization."""
        return {
            "game_title": self.game_title,
            "config_path": self.config_path,
            "menu_options": [option.to_dict() for option in self.menu_options],
            "categories": self.categories,
            "constraints": add_order_key(self.constraints),
        }


def add_order_key(data: dict[str, Any]) -> dict[str, Any]:
    """Adds an 'Order' key to maintain key ordering in the Wails App.

    Go does not maintain order for maps.
    """
    if isinstance(data, dict):
        ordered_keys = list(data.keys())

        order = [key for key in ordered_keys if key and key[0].isupper()]

        if order:
            data["Order"] = order

        for key in ordered_keys:
            if isinstance(data[key], dict):
                add_order_key(data[key])

    return data

File: internal/ipc/game_gui.go
package ipc

type MenuOption struct {
	Label string   `json:"label"`
	Args  []string `json:"args"`
	Order int      `json:"order"`
}

type GameGUI struct {
	GameTitle   string                 `json:"game_title"`
	ConfigPath  string                 `json:"config_path"`
	MenuOptions []MenuOption           `json:"menu_options"`
	Constraints map[string]interface{} `json:"constraints"`
	Order       int                    `json:"order"`
}
